src/cmlextensions/ray_cluster/ray_cluster.py

```diff
@@ -11,72 +11,105 @@
 # use of the file.
 
 import os
-import cdsw
+import inspect
+
+# PBJ Runtimes do not have the cdsw library installed.
+# Instead, the cml library is added to workloads in recent CML releases.
+try:
+    import cml.utils_v1 as utils
+    cdsw = utils._emulate_cdsw()
+except ImportError:
+    import cdsw
+
 
 DEFAULT_DASHBOARD_PORT = os.environ['CDSW_APP_PORT']
 
 class RayCluster():
     """Ray Cluster built on CML Worker infrastructure"""
 
-    def __init__(self, num_workers, worker_cpu=2, worker_memory=4, head_cpu=2, head_memory=4, dashboard_port=DEFAULT_DASHBOARD_PORT, env
+    def __init__(self, num_workers, worker_cpu=2, worker_memory=4, worker_nvidia_gpu=0, head_cpu=2, head_memory=4, head_nvidia_gpu=0,  dashboard_port=DEFAULT_DASHBOARD_PORT, env
 ={}):
         self.num_workers = num_workers
         self.worker_cpu = worker_cpu
         self.worker_memory = worker_memory
+        self.worker_nvidia_gpu = worker_nvidia_gpu
         self.head_cpu = head_cpu
         self.head_memory = head_memory
+        self.head_nvidia_gpu = head_nvidia_gpu
         self.dashboard_port = dashboard_port
         self.env = env
 
         self.ray_head_details = None
         self.ray_worker_details = None
 
+    def _stop_ray_workloads(self):
+        for workload_details in [self.ray_head_details, self.ray_worker_details]:
+            for status in ["workers" , "failures"]:
+                if workload_details is not None and status in workload_details:
+                    stop_responses = cdsw.stop_workers(*[workload["id"] for workload in workload_details[status]])
+                    stop_response_statuses = [response.status_code for response in stop_responses]
+                    if(any([status >= 300 for status in stop_response_statuses])):
+                        print("Could not stop all Ray workloads. Trying to force top all CML workers created in this session.")
+                        cdsw.stop_workers()
+
+    def _start_ray_workload(self, args, startup_timeout_seconds):
+        workloads =  cdsw.launch_workers(**args)
+        workloads_with_ids = [wl for wl in workloads if "id" in wl]
+        workloads_with_no_ids = [wl for wl in workloads if "id" not in wl]
+        no_id_messages = set([wl.get("message", "") for wl in workloads_with_no_ids])
+        if len(workloads_with_no_ids) > 0:
```

Comment:
Don't we need to do this after the await_workers call so we wait till the timeout? Is this a failure mode that's unrelated to timeouts?